pkg/reconciler/stabilizing/stabilizing_controller.go

```diff
@@ -0,0 +1,422 @@
+
+	rgObj := &v1alpha1.ResourceGroup{}
+	if err := r.client.Get(ctx, req.NamespacedName, rgObj); err != nil {
+		switch {
+		case meta.IsNoMatchError(err):
+			return reconcile.Result{},
+				fmt.Errorf("ResourceGroup CRD not installed or established: %w", err)
+		case apierrors.IsNotFound(err):
+			r.Logger(ctx).V(1).Info("Skipping Stabilizing condition update: ResourceGroup object not found")
+			return reconcile.Result{}, nil
+		default:
+			return reconcile.Result{},
+				fmt.Errorf("getting ResourceGroup object from cache: %s: %w",
+					req.NamespacedName, err)
+		}
+	}
+
+	if rgObj.Status.ObservedGeneration != rgObj.Generation {
+		r.Logger(ctx).V(1).Info("Skipping Stabilizing condition update: ResourceGroup status update pending")
+		return reconcile.Result{}, nil
+	}
+	if rgObj.DeletionTimestamp != nil {
+		r.Logger(ctx).V(1).Info("Skipping Stabilizing condition update: ResourceGroup being deleted")
+		return reconcile.Result{}, nil
+	}
+
+	stabilizingCondition, err := computeStabilizingCondition(rgObj)
+	if err != nil {
+		return reconcile.Result{},
+			fmt.Errorf("failed to compute stabilizing status: %w", err)
+	}
+
+	if err := r.applyStabilizingCondition(ctx, stabilizingCondition); err != nil {
+		return reconcile.Result{},
+			fmt.Errorf("failed to apply stabilizing status condition on object: %v, %w",
+				r.syncID, err)
+	}
+
+	r.Logger(ctx).V(5).Info("Reconcile completed")
+	return reconcile.Result{}, nil
+}
+
+func computeStabilizingCondition(rgObj *v1alpha1.ResourceGroup) (SyncCondition, error) {
+	actuationStatusCounts := map[v1alpha1.Actuation]int{
+		v1alpha1.ActuationFailed:    0,
+		v1alpha1.ActuationSkipped:   0,
+		v1alpha1.ActuationPending:   0,
+		v1alpha1.ActuationSucceeded: 0,
+	}
+	reconcileStatusCounts := map[v1alpha1.Reconcile]int{
+		v1alpha1.ReconcileFailed:    0,
+		v1alpha1.ReconcileSkipped:   0,
+		v1alpha1.ReconcilePending:   0,
+		v1alpha1.ReconcileSucceeded: 0,
+	}
+
+	for i, objStatus := range rgObj.Status.ResourceStatuses {
+		switch objStatus.Actuation {
+		case v1alpha1.ActuationFailed, v1alpha1.ActuationSkipped, v1alpha1.ActuationPending, v1alpha1.ActuationSucceeded:
+			actuationStatusCounts[objStatus.Actuation]++
+		default:
+			return SyncCondition{},
+				fmt.Errorf("invalid ResourceGroup field value: status.resourceStatuses[%d].actuation: `%s`",
+					i, objStatus.Actuation)
+		}
+		switch objStatus.Reconcile {
+		case v1alpha1.ReconcileFailed, v1alpha1.ReconcileSkipped, v1alpha1.ReconcilePending, v1alpha1.ReconcileSucceeded:
+			reconcileStatusCounts[objStatus.Reconcile]++
+		default:
+			return SyncCondition{},
+				fmt.Errorf("invalid ResourceGroup field value: status.resourceStatuses[%d].actuation: `%s`",
+					i, objStatus.Actuation)
+		}
+	}
+
+	// Get the source commit from the inventory
+	commit := core.GetAnnotation(rgObj, metadata.SourceCommitAnnotationKey)
+
+	syncCondition := SyncCondition{
+		Type:   SyncStabilizing,
+		Commit: commit,
+	}
+
+	switch {
```

Comment:
Will it be a case when multiple cases are true at the same time, causing the syncCondition being overwritten?